srml/staking/src/lib.rs

```diff
@@ -1017,30 +1033,61 @@ impl<T: Trait> Module<T> {
 	/// get a chance to set their session keys.
 	fn new_era() -> Option<Vec<T::AccountId>> {
 		// Payout
-		let reward = <CurrentEraReward<T>>::take();
-		if !reward.is_zero() {
+		let rewards = <CurrentEraRewards<T>>::take();
+		let now = <timestamp::Module<T>>::now();
+		let previous_era_start = <CurrentEraStart<T>>::mutate(|v| {
+			rstd::mem::replace(v, now.clone())
```

Comment:
If I have understood your code correctly here: the goal here is to perform for a storage value: `::put()` and `::take()` at the same time. Maybe this should be abstracted on the storage layer as it looks like a common pattern rather than touching `mem` directly.

Reply:
having a replace method on the trait makes sense.

Also even if mem modules looks touchy actually mem::replace is a very simple and safe function